### ISSUE TITLE
Add check for unused private fields

### ISSUE THREAD

**Issue by KostyaSha** [checkstyle/checkstyle#12596]
**Is your feature request related to a problem? Please describe.**
Like `UnusedLocalVariable` it would be good to have unused private field check.

**Describe the solution you'd like**
Probably new checker.

**Additional context**
https://github.com/checkstyle/checkstyle/pull/10930

**Comment by nrmancuso** [comment]
Please follow issue template with code examples, etc.

**Comment by KostyaSha** [comment]
> issue template

I used feature request template.

> with code examples

Added

**Comment by romani** [comment]
Please read and follow https://checkstyle.org/report_issue.html#How_to_request_new_Check.2FModule.3F if you want to speed up approval of this issue

**Comment by nuk6** [comment]
Is there a way to escalate this?
This is one of the very important checks!

**Comment by romani** [comment]
@KostyaSha , are you ok to rename to UnusedPrivateClassField ?

As we plan to target only private fields.

**Comment by nrmancuso** [comment]
> UnusedPrivateClassField

This sounds more like a check for private classes, I would prefer something like `UnusedPrivateField`.

**Comment by commodis** [comment]
Would it be useful to extend this check for `static` variables? Private constants suffer from the same problem.
I could open a new issue in case it should be handled separately.

**Comment by rnveach** [comment]
I would first work on initial instance of module to see what is first required for the basic functionality of instance fields, and then determine if adding support for static fields is similar in terms of the code needed and tracking. If it is not similar, which I am assuming this to be the case, then I would recommend it being 2 different checks. It also makes sense for it to be 2 different checks in terms of making it easier on users to enable and disable the specifics they want.

**Comment by romani** [comment]
I am ok with this design, we can cover static too because if it is private and not used, then static/final/.....any extra modifiers... does not matter.

We need to add property `ignoreAnnotatedBy` and `ignoreNestingTypeAnnotatedBy` to accommodate modern trends of Lombok annotations https://projectlombok.org/features/Data

What about java record ? https://docs.oracle.com/en/java/javase/17/language/records.html
Should we put `Class` word in name of Check to make it clear target is only `class`, so `enum` unused fields will be ignored.

@rnveach , @nrmancuso , @KostyaSha , @nuk6 , @commodis , please share your ideas.

**Comment by rnveach** [comment]
> Should we put Class word in name of Check to make it clear target is only class, so enum unused fields will be ignored.

It shouldn't matter if it is class or enum. Private fields that aren't used are not used. Enum fields are public and used, so they wouldn't apply to this check's logic.
Example: https://gist.github.com/schwarzeszeux/2347102

**Comment by romani** [comment]
@rnveach , ok, are you ok with 2  extra properties and design of Check?

**Comment by rnveach** [comment]
I think the issue should stick the basics of just implementing private field support and have another issue add other properties. If there is some delay in other issue, xpath suppression is always an option.

**Comment by romani** [comment]
I am ok with splitting and implementing extra properties as separate issue/PRs.

@nrmancuso , please review and approve issue.

**Comment by nrmancuso** [comment]
I am good with proposal at https://github.com/checkstyle/checkstyle/issues/12596#issuecomment-1831163528, namely:
1. Create basic check, with no additional properties 
2. Create issue as we develop check to review the need for additional properties that may come up
3. Have at least one example in check documentation with Lombok/Spring json property annotation suppression by xpath

**Comment by sktpy** [comment]
I am on it.

**Comment by sktpy** [comment]
Was done with this check but made a _huge_ mistake while designing this as I had a wrong invariant in mind, that being the private instance fields cannot be accessed outside its defining block, but this was only true for top level classes. ;) :man_facepalming: 

```java
class TopLevel1 {
    private int privateField;

    class Inner1 {
        private int privateFieldOrIsIt;
    }

    class Inner2 {
        Inner1 inner = new Inner1();
        int iSpy = inner.privateFieldOrIsIt; // pain
    }
}

class TopLevel2 {
    TopLevel1 topLevel = new TopLevel1();
    int iCannotSpy = topLevel.privateField; // compilation error
}
```

On the redesign I go :runner:
![image](https://github.com/checkstyle/checkstyle/assets/145206423/505a08b0-410b-4a20-8913-9b091c4c8813)

**Comment by sktpy** [comment]
I will visit this later as I am running low on time atm, but if someone else would want to work on it before I am able to push the PR, feel free to take this issue.

**Comment by Abdelrhmansersawy** [comment]
@romani 
I'm starting to work on this new check. 
To better understand this type of check design, I will review similar unused checks such as `UnusedLocalVariable`, `UnusedVariable`, and `UnusedImports`.  
After that, I will open a draft PR to discuss my design.

**Comment by roded** [comment]
We're looking forward to this check. Any updates on it since April? Thanks

**Comment by Abdelrhmansersawy** [comment]
> We're looking forward to this check. Any updates on it since April? Thanks

No, I’m not working on this check.

**Comment by ZaheerAhmadDev** [comment]
@romani,
I’m planning to work on this issue and wanted to confirm it’s still relevant.
My idea is to start with a basic `UnusedPrivateFieldCheck` and use one combined example that demonstrates:
provide one combined example covering unused fields, assigned-only fields, static fields, and inner-class usage.
If this still matches the intended direction i will open a draft PR.
Please let me know if you’d prefer a different scope or example structure.

**Comment by romani** [comment]
@ZaheerAhmadDev, please help community to get this validation.
You are good to start.

**Comment by avadhutmali** [comment]
I am on it

**Comment by romani** [comment]
there is already https://github.com/checkstyle/checkstyle/pull/19816, we hope it will be merged.

**Comment by romani** [comment]
Check is merged !
all gratitude to @ZaheerAhmadDev .
